packages/metro-file-map/src/lib/TreeFS.js

```diff
@@ -0,0 +1,386 @@
+  MutableFileSystem,
+  Path,
+} from '../flow-types';
+
+import H from '../constants';
+import * as fastPath from '../lib/fast_path';
+import invariant from 'invariant';
+import path from 'path';
+
+type DirectoryNode = Map<string, AnyNode>;
+type FileNode = FileMetaData;
+type LinkNode = string;
+type AnyNode = FileNode | DirectoryNode | LinkNode;
+
+export default class TreeFS implements MutableFileSystem {
+  +#rootDir: Path;
+  +#files: FileData;
+  +#rootNode: DirectoryNode = new Map();
+
+  constructor({rootDir, files}: {rootDir: Path, files: FileData}) {
+    this.#rootDir = rootDir;
+    this.#files = files;
+    this.bulkAddOrModify(files);
+  }
+
+  getSerializableSnapshot(): FileData {
+    return new Map(
+      Array.from(this.#files.entries(), ([k, v]: [Path, FileMetaData]) => [
+        k,
+        [...v],
+      ]),
+    );
+  }
+
+  getModuleName(file: Path): ?string {
+    const fileMetadata = this._getFileData(file);
+    return (fileMetadata && fileMetadata[H.ID]) ?? null;
+  }
+
+  getSize(file: Path): ?number {
+    const fileMetadata = this._getFileData(file);
+    return (fileMetadata && fileMetadata[H.SIZE]) ?? null;
+  }
+
+  getDependencies(file: Path): ?Array<string> {
+    const fileMetadata = this._getFileData(file);
+
+    if (fileMetadata) {
+      return fileMetadata[H.DEPENDENCIES]
+        ? fileMetadata[H.DEPENDENCIES].split(H.DEPENDENCY_DELIM)
+        : [];
+    } else {
+      return null;
+    }
+  }
+
+  getSha1(file: Path): ?string {
+    const fileMetadata = this._getFileData(file);
+    return (fileMetadata && fileMetadata[H.SHA1]) ?? null;
+  }
+
+  exists(file: Path): boolean {
+    const result = this._getFileData(file);
+    return result != null;
+  }
+
+  getAllFiles(): Array<Path> {
+    return Array.from(this._regularFileIterator(), normalPath =>
+      this._normalToAbsolutePath(normalPath),
+    );
+  }
+
+  linkStats(file: Path): ?FileStats {
+    const fileMetadata = this._getFileData(file, {follow: false});
+    if (fileMetadata == null) {
+      return null;
+    }
+    const fileType = fileMetadata[H.SYMLINK] === 0 ? 'f' : 'l';
+    const modifiedTime = fileMetadata[H.MTIME];
+    invariant(
+      typeof modifiedTime === 'number',
+      'File in TreeFS missing modified time',
+    );
+    return {
+      fileType,
+      modifiedTime,
+    };
+  }
+
+  matchFiles(pattern: RegExp | string): Array<Path> {
+    const regexpPattern =
+      pattern instanceof RegExp ? pattern : new RegExp(pattern);
+    const files = [];
+    for (const filePath of this._pathIterator()) {
+      const absolutePath = this._normalToAbsolutePath(filePath);
+      if (regexpPattern.test(absolutePath)) {
+        files.push(absolutePath);
+      }
+    }
+    return files;
+  }
+
+  /**
+   * Given a search context, return a list of file paths matching the query.
+   * The query matches against normalized paths which start with `./`,
+   * for example: `a/b.js` -> `./a/b.js`
+   */
+  matchFilesWithContext(
```

Comment:
FYI @EvanBacon - this is an O(subtree) implementation of `matchFilesWithContext`, so once TreeFS is the default (a bit of benchmarking + optimisation to do) we'll have addressed [this](https://github.com/facebook/metro/blob/a0ba9945ea0bcb99e2048eb76dafafa03a0e1709/packages/metro/src/lib/transformHelpers.js#L154) and can probably promote `require.context` to stable 🚀